posts/2017-07-02-databases.md

```diff
@@ -0,0 +1,90 @@
+soundcloud: 330991445
+author: Dave Taylor
+---
+
+
+## Databases
+
+Interactions with a database can be expressed clearly and succinctly in a functional paradigm. Developers coming to functional programing from an object oriented background, may feel like something is simply missing and search for an abstraction. 
+
+The functional approach is to simply make a connection and creating functions. By using a DSL combined with the expressive power of compound procedures, apply and piping a programmer can express his interactions with the database. These expressions are often both readable and intuitive, to other developers. 
+
+
+### The Object Oriented Approach
+
+Looking at some alternative Java approaches (which have been greatly simplified), we can begin to see the scaffolding needed to maintain these interactions. A side effect of these object oriented abstractions is code bloat. This implementation detail introduces the tendency to have the code drive the design of the database, often when not appropriate. 
+
+ 
+#### Hibernate: 
+
+This [popular paradigm](http://docs.jboss.org/hibernate/orm/5.2/quickstart/html_single/) requires a programmer to create a class with meta data before querying. Keeping this logic in sync with your database migrations can be tedious. 
+
+
+```
+@Entity
+@Table(name = "BOOK")
+public class BOOK {
+
+  @Id
+  @Column(name = "id")
+  private long id;
+
+  @Column(name = "name")
+  private String name;
+
+  @Column(name = "author")
+  private String author;
+```
+
+```
+Query query = session.createQuery("from BOOK");
+    List<Car> empList = query.list();
+```
+
+#### JDBI
+
+
+The JDBI [library](http://jdbi.org/sql_object_api_queries/) was designed to map queries to an interface. Like Hibernate a developer needs to create a custom class implementing ResultSetMapper to map each element in a database to a class. 
+
+```
+public interface Queries
+{
+  @SqlQuery("select name from BOOK where id = :id")
+  String findName(@Bind("id") int id);
+
+  @SqlQuery("select name from BOOK where id > :from and id < :to")
+  List<String> findNamesBetween(@Bind("from") int from, @Bind("to") int to);
+
+  @SqlQuery("select name from BOOK")
+  Iterator<String> findAllNames();
+}
+```
+
+
+
+### The Functional Approach
+
+Looking at the Clojure approach with [HoneySQL](https://github.com/jkk/honeysql), the absence of code bloat from the OO paradigm is noticeable and the functions are readable. Moving the focus to functions and away from custom types introduces simplicity. 
+
+
+```
+(def books-query (-> (select :*)
+                     (from [:BOOKS :B])))
+
+(def books-query-by-author
+  [author] (-> books-query 
+         (merge-where
+                [:= :B.author author]))
+
+(def books-query-by-published-year
+  [year] (-> books-query
+        (merge-where
+                [:= :B.year year]))
+
+(def books-query-by (comp books-query merge-where))
+```
```

Comment:
Not sure if it is worth it, but maybe showing how the actual query is run... using `jdbc`.... maybe that would make it a bit more clear as well ...